### PR TITLE
feat: provide a consistent PersiaBatch interface

### PR DESCRIPTION
**Pull request by williamstar**
- Support multiple dtype for `not_id_type_tensors` and `labels` 
- Support dynamic shape of `not_id_type_tensors` and `labels`

### CODE REVIEW

File: e2e/adult_income/data_loader.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 from tqdm import tqdm
 
-from persia.prelude import PersiaBatch
+from persia.embedding.data import PersiaBatch
```
Comment (snowpeakz) [inline e2e/adult_income/data_loader.py:5]: As we have `non_id_type_feature` in `PersiaBatch`, it seems better to import from `persia.data` or `persia.dataflow` .

File: e2e/adult_income/data_loader.py [hunk 2]
```diff
@@ -19,9 +19,10 @@
...
-            batch_data.add_non_id_type_feature([dense])
-            batch_data.add_id_type_features(batch_sparse_ids)
-            batch_data.add_label(target)
-            ctx.send_data(batch_data)
+        for (non_id_type_feature, id_type_features, label) in tqdm(
+            loader, desc="gen batch data..."
+        ):
+            persia_batch = PersiaBatch(id_type_features)
+            persia_batch.add_non_id_type_feature(non_id_type_feature)
+            persia_batch.add_label(label)
```
Comment (snowpeakz) [inline e2e/adult_income/data_loader.py:27]: If we have python class `PersiaIdTypefeatures`, and implement `PersiaIdTypefeatures.from_nparray_list` for this class, then call `persia_batch.add_id_type_feature(id_type_feature)`, would it easier for users to learn what they should pass.

Comment (snowpeakz) [reply e2e/adult_income/data_loader.py:27]: The same for `PersiaNonTypefeatures ` and `PersiaLabel`

File: persia/embedding/data.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+
+class PersiaBatch:
+    r"""PersiaBatch store the id_type_features, non_id_type_features, labels。
+
+    Example:
+        >>> import time
+        >>> import json
+        >>> import numpy as np
+        >>> from persia.data import PersiaBatch
```
Comment (snowpeakz) [inline persia/embedding/data.py:26]: Need to be consistent with latest code.

File: e2e/adult_income/serve_client.py [hunk 4]
```diff
@@ -46,11 +47,11 @@ def infer(stub, model_name, model_input):
...
-    for (dense, batch_sparse_ids, target) in tqdm(loader, desc="gen batch data..."):
-        batch_data = PersiaBatch()
-        batch_data.add_non_id_type_feature([dense])
-        batch_data.add_id_type_features(batch_sparse_ids, False)
-
+    for (non_id_type_feature, id_type_features, target) in tqdm(
+        loader, desc="gen batch data..."
+    ):
+        batch_data = PersiaBatch(id_type_features, requires_grad=False)
+        batch_data.add_non_id_type_feature(non_id_type_feature)
```
Comment (snowpeakz) [inline e2e/adult_income/serve_client.py:54]: The naming of `non_id_type_feature` may cause confusion. For example, we have 5 dense features in adult income dataset, while in our data procession, we concat them and regarded them as 1 dense feature.
How about use `add_concat_non_id_type_features` ?

File: e2e/adult_income/data_loader.py (hunk 1 above)
Comment (williamstar) [reply e2e/adult_income/data_loader.py:5]: Due `non_id_type_feature` is `optional`, I think the `PersiaBatch` is not suitable to place into `persia.data`.And the major functionality of `persia.data` will focus on procssing `dataset` and `dataloder`.

File: e2e/adult_income/serve_client.py (hunk 4 above)
Comment (williamstar) [reply e2e/adult_income/serve_client.py:54]: `non_id_type_feature` is a tensor with ndim greater than 1.

File: e2e/adult_income/data_loader.py (hunk 1 above)
Comment (snowpeakz) [reply e2e/adult_income/data_loader.py:5]: I am a little worried that users mistakenly think that we are mapping non id type features to embeddings through some method, if he didn’t read the document carefully. Maybe this worry is redundant.

File: e2e/adult_income/data_loader.py (hunk 2 above)
Comment (snowpeakz) [reply e2e/adult_income/data_loader.py:27]: After offline discussion, we will do it later. https://github.com/PersiaML/PERSIA/issues/183

**Review by snowpeakz — APPROVED** [review]
It is better to create an issue about https://github.com/PersiaML/PERSIA/pull/181#discussion_r756618027 as we discussed. @williamstar

File: e2e/adult_income/data_loader.py (hunk 1 above)
Comment (snowpeakz) [reply e2e/adult_income/data_loader.py:5]: I think we need an issue to remind us to refactor this.

**Review by NOBLES5E — APPROVED** [review]

File: e2e/adult_income/data_loader.py [hunk 5]
```diff
@@ -19,9 +19,10 @@
...
     with DataCtx() as ctx:
         _, loader = make_dataloader(train_filepath)
-        for (dense, batch_sparse_ids, target) in tqdm(loader, desc="gen batch data..."):
-            batch_data = PersiaBatch()
-            batch_data.add_non_id_type_feature([dense])
-            batch_data.add_id_type_features(batch_sparse_ids)
-            batch_data.add_label(target)
-            ctx.send_data(batch_data)
+        for (non_id_type_feature, id_type_features, label) in tqdm(
+            loader, desc="gen batch data..."
```
Comment (snowpeakz) [inline e2e/adult_income/data_loader.py:23]: `generating`

**Review by snowpeakz — CHANGES_REQUESTED** [review]
See comment

**Review by snowpeakz — APPROVED** [review]

File: e2e/adult_income/data_loader.py (hunk 1 above)
Comment (williamstar) [reply e2e/adult_income/data_loader.py:5]: Related to #185
